channeldb/reports.go

```diff
@@ -0,0 +1,332 @@
+package channeldb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+var (
+	// reportBucket is a bucket which holds a record of the final set
+	// of resolutions which were processed for a channel.
+	// [reportBucket]
+	//	[chainHashBucket]
+	//		[channelBucket]
+	//			[resolversBucket]
+	reportBucket = []byte("channel-reports")
+
+	// resolversBucket holds the outcome of a channel's resolvers.
+	resolversBucket = []byte("resolvers-bucket")
+)
+
+var (
+	// ErrNoReportBucket is returned if the report bucket has not been
+	// created yet.
+	ErrNoReportBucket = errors.New("no reports bucket present")
+
+	// ErrNoChainHashBucket is returned when we have not created a bucket
+	// for the current chain hash.
+	ErrNoChainHashBucket = errors.New("no chain hash bucket")
+
+	// ErrNoChannel is returned when a channel is not found in the chain
+	// hash bucket.
+	ErrNoChannel = errors.New("channel bucket not found")
+)
+
+// ResolverOutcome describes the outcome for the resolver that the contract
+// court reached. This state is not necessarily final, htlcs on our own
+// commitment are resolved across two resolvers.
+type ResolverOutcome uint8
+
+const (
+	resolverOutcomeUnknown ResolverOutcome = iota
+
+	// ResolverOutcomeAnchorRecovered indicates that we swept our own
+	// anchor.
+	ResolverOutcomeAnchorRecovered
+
+	// ResolverOutcomeAnchorLost indicates that we did not sweep our anchor,
+	// or another party claimed it.
+	ResolverOutcomeAnchorLost
+
+	// ResolverOutcomeAnchorUnclaimed indicates that the anchor was left
+	// unspent.
+	ResolverOutcomeAnchorUnclaimed
+
+	// ResolverOutcomeOutgoingHtlcTimeout indicates the we successfully
+	// timed out one of our outgoing htlcs on chain.
+	ResolverOutcomeOutgoingHtlcTimeout
+
+	// ResolverOutcomeOutgoingHtlcClaim indicates that the remote party
+	// claimed one of our outgoing htlcs on chain with the preimage.
+	ResolverOutcomeOutgoingHtlcClaim
+
+	// ResolverOutcomeIncomingHtlcClaimed indicates that we claimed an
+	// incoming htlc with its preimage on chain.
+	ResolverOutcomeIncomingHtlcClaimed
+
+	// ResolverOutcomeIncomingHtlcSuccessTx indicates that we broadcast the
+	// first stage of a two stage htlc claim for an incoming htlc.
+	ResolverOutcomeIncomingHtlcSuccessTx
```

Comment:
This isn't a final outcome. Will that ever be stored?

Reply:
As is, it's stored when we hand off to a success resolver. I think it's important to have this intermediate state recorded on disk, so that we have a full record of all on chain activity.